Added JumpStepWrapper

andyElking opened the pull request:
Hi Patrick,

I factored the `jump_ts` and `step_ts` out of the `PIDController` into `JumpStepWrapper` (I'm not very set on this name, lmk if you have ideas). I also made it behave as we discussed in #483. In particular, the following three rules are maintained:

1. We always have `t1-t0 <= prev_dt` (this is checked via `eqx.error_if`), with inequality only if the step was clipped or if we hit the end of the integration interval (we do not explicitly check for that).
2. If the step was accepted, then `next_dt` must be `>=prev_dt`.
3. If the step was rejected, then `next_dt` must be `< t1-t0`.

We achieve this in a very simple way here: https://github.com/patrick-kidger/diffrax/blob/78b122adf39b2f8d26a79d0ac239a2fb675653a1/diffrax/_step_size_controller/jump_step_wrapper.py#L119-L123

The next step is to add a parameter `JumpStepWrapper.revisit_rejected_steps` which does what you expect. That will appear in a future commit in this same PR.

andyElking commented:
I now also added the functionality to revisit rejected steps. In addition, I also imporved the runtime of `step_ts` and `jump_ts`, because the controller no longer searches the whole array each time, but keeps an index of where in the array it was previously.

Also I think there was a bug in the PID controller, where it would sometimes reject a step, but have `factor>1`. To remedy this I modified the following:
https://github.com/patrick-kidger/diffrax/blob/501bed52fb3726aaee41798aaab46d2bdff1a680/diffrax/_step_size_controller/adaptive.py#L569-L574
I think possibly something smaller than just `self.safety` would make even more sense, I feel like if a step is rejected the next step should be at least 0.5x smaller. But I'm not an expert.

I added a test for revisiting steps and it all seems to work. I also sprinkled in a bunch of `eqx.error_if` statements to make sure the necessary invariants are always maintained. But this is a bit experimental, so maybe there are some bugs I didn't test for.

I think I commented the code quite well, so hopefully you can easily notice if I made a mistake somewhere.

P.S.: Sorry for bombarding you with PRs. As far as I'm concerned this one is very low priority, I can use the code even if it isn't merged into diffrax proper.

andyElking commented:
Hi @patrick-kidger,
I got rid of some `eqx.error_if`s that I added to my `JumpStepWrapper` and redid the timing benchmarks. My new implementation was already faster than the old `PIDController` before, but now this is way more significant, especially when `step_ts` is long (think `>100`). Surprisingly, it is faster even when it has to revisit rejected steps. See https://github.com/patrick-kidger/diffrax/blob/345e23ad34abe095d25b8be8d621360afd592165/benchmarks/jump_step_timing.py#L126-L128

patrick-kidger commented on diffrax/_step_size_controller/adaptive.py:
```diff
@@ -353,35 +349,14 @@ def dynamics(t, y, args):
     dtmin: Optional[RealScalarLike] = None
     dtmax: Optional[RealScalarLike] = None
     force_dtmin: bool = True
-    step_ts: Optional[Real[Array, " steps"]] = eqx.field(
-        default=None, converter=_none_or_array
-    )
-    jump_ts: Optional[Real[Array, " jumps"]] = eqx.field(
-        default=None, converter=_none_or_array
-    )
```
So I think it's important that we don't break the existing `PIDController`. If at all possible I really try to avoid backward-incompatible changes.

In this case I think we have two choices:
1. Keep all of the code here, but if `self.step_ts is not None` or `self.jump_ts is not None`, then emit a warning in `__check_init__` to prefer the new implementation instead.
2. Add a custom metaclass with a custom `__call__`, so that if `step_ts` or `jump_ts` is passed, then in fact we'll automatically rewrap them into the new implementation. (That is, `PIDController(..., jump_ts=...)` will actually return a `JumpStepWrapper(PIDController(...), ...)` ! Python allows for some crazy things.)

andyElking replied:
Yes, I was just about to ask your advice on how to achieve the second point 😁.
Will do!

patrick-kidger commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,410 @@
+# If `i_rjct < 0` then an error is raised.
+
+
+class JumpStepWrapper(
+    AbstractStepSizeController[_JumpStepState[_ControllerState], _Dt0]
+):
```
I think we should make this -- and the wrapped `.controller` -- specifically be `AbstractAdaptiveStepSizeController`. We do have a couple of places that check for adaptivity via `isinstance(...., AbstractAdaptiveStepSizeController)`.

(More generally this is something where in retrospect I probably shouldn't have lifted certain properties, like adaptivity, into the type system, as it makes wrapping harder. At least in this case I think it's fine, I don't really anticipate use cases for this with fixed step size controllers -- do you?)

andyElking replied:
I agree, probably making it adaptive-only makes sense.

patrick-kidger commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,410 @@
+        self.controller = controller
+        self.step_ts = _none_or_array(step_ts)
+        self.jump_ts = _none_or_array(jump_ts)
+        self.rejected_step_buffer_len = rejected_step_buffer_len
```
Can we default this to `max_steps`, do you think? (Maybe it is set as `None`?)

andyElking replied:
In practice it rarely needs to be more than 10, so `max_steps` would be way overkill in my opinion. But I agree, using `None` to turn it off is more transparent.

patrick-kidger commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,410 @@
+        self.jump_ts = _none_or_array(jump_ts)
+        self.rejected_step_buffer_len = rejected_step_buffer_len
+        self.callback_on_reject = _callback_on_reject
+        self.__check_init__()
```
This isn't necessary, Equinox will call this for you.

andyElking replied:
Oh, thanks. This was actually added by copilot and I wasn't sure, so I left it 😅.

patrick-kidger commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,410 @@
+                f"jump_ts must be floating point, not {self.jump_ts.dtype}"
+            )
+
+    def wrap(self, direction: IntScalarLike):
+        step_ts = None if self.step_ts is None else self.step_ts * direction
+        jump_ts = None if self.jump_ts is None else self.jump_ts * direction
+        return eqx.tree_at(
+            lambda s: (s.step_ts, s.jump_ts),
+            self,
+            (step_ts, jump_ts),
+            is_leaf=lambda x: x is None,
+        )
```
We should call `.wrap` of the controller we are wrapping too, to give it an opportunity to do anything else it wants to do.

andyElking replied:
Oh yes of course, completely forgot about that.

patrick-kidger commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,410 @@
+            step_ts = upcast_or_raise(
+                self.step_ts,
+                jnp.zeros((), tdtype),
+                "`PIDController.step_ts`",
```
We should change the name `PIDController` here.

andyElking replied:
Haha oops

patrick-kidger commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,410 @@
+        else:
+            step_ts = None
+
+        if self.jump_ts is not None:
```
Nit: remove the `not` and flip the branches. We can save four characters ;)

(By which I mean, I find the logic easier to follow.)

patrick-kidger commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,410 @@
+            jump_ts = None
+
+        if self.rejected_step_buffer_len > 0:
+            rjct_buff = jnp.zeros(
```
Can we spell this out? I try to avoid less-readable abbreviated variable names.

andyElking replied:
Fair!

patrick-kidger commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,410 @@
+            # If the step was rejected, then we need to store the rejected time in the
+            # rejected buffer and decrement the rejected index.
+            i_rjct = jnp.where(keep_step, i_rjct, i_rjct - 1)
+            i_rjct = eqx.error_if(
+                i_rjct,
+                i_rjct < 0,
+                "Maximum number of rejected steps reached. "
+                "Consider increasing JumpStepWrapper.rejected_step_buffer_len.",
+            )
```
And in particular we should be able to skip this check if we set this value to `max_steps`.

andyElking replied:
I'll run some timing tests to see if a long buffer significantly hinders performance. If not, I'm happy to change it to `max_steps`.

andyElking replied:
So it seems that using a long buffer doesn't impact performance at all. So I think it could be fine to just make a binary parameter `revisit_rejected` and if it is `True`, then the buffer length is `max_steps`. However when vmaping over this that could create a significant but unnecessary memory overhead, so maybe giving the option to change this buffer length might be reasonable. What do you think?

patrick-kidger replied:
Ah, good point about the vmap case.

I think the parameter `revisit_rejected` could be of type `bool | int`. If `False` then don't revisit. If `True` then do revisit and use a buffer of size `max_steps`. If an integer then do revisit and use a buffer of that size. Gives an easy choice whilst also offering control if that's really necessary for some reason.

patrick-kidger commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,410 @@
+                "Maximum number of rejected steps reached. "
+                "Consider increasing JumpStepWrapper.rejected_step_buffer_len.",
+            )
+            rjct_buff = jnp.where(keep_step, rjct_buff, rjct_buff.at[i_rjct].set(t1))
```
Note that this is a very expensive way to describe this operation! You're copying the whole buffer. XLA will *sometimes* optimize this out -- because I added that optimization to it! -- but not always.

Better is to do `rjct_buff.at[i_rjct].set(jnp.where(keep_step, rjct_buff[i_rjct], t1))`

patrick-kidger replied:
Other than that, I think we may need to extend the API here slightly -- we should be able to mark state like this as being a buffer for the purposes of:

https://github.com/patrick-kidger/diffrax/blob/0679807bf7bda0de3884a78cf20ce2dd4d8ade4c/diffrax/_integrate.py#L621

which is needed to avoid spurious copies during backpropagation.

(You can see that both of these comments are basically us having to work around limitations of the XLA compiler.)

andyElking replied:
The first one makes sense, I should have seen this.

I don't really know what you want me to do in your second comment. And frankly `diffeqsolve` is something I haven't even started digesting yet. Are you telling me `rejected_buffer` should be one of the `outer_buffers`, meaning that I should make it an instance of `SaveState` or sth like that? I would apprecaite a bit more guidance.

Also damn how you managed to write all this code is beyond me. Even trying to begin understanding it seems a lot!

patrick-kidger replied:
Haha, you're too kind!

As for my second comment -- I think I've realised that I was wrong. Let me explain. Our backpropagation involves saving copies of our state in checkpoints. Let's suppose we set `RecursiveCheckpointAdjoint(checkpoints=max_steps)`, so that's `O(max_steps)` memory right? Well, not quite: our updating buffer here is potentially of length `max_steps` (as per the debate above), and we're saving a copy of it in every checkpont, so we'd actually be using `O(max_steps^2)` memory!

The simple solution to this will just be to set the size of this buffer to e.g. 100 by default, and just allow those copies to be made. And given the behaviour you have here -- in which you potentially overwrite values -- then that is actually what's necessary as well.

As for the complicated solution that I was wrong about: let's consider the case of `SaveAt(steps=True)`. This also involves a buffer of length `max_steps`, that we save into as we go along. Fortunately, this one has a useful extra property, which is that we never overwrite a value. That means we don't actually need to copy our buffer for every checkpoint! We can use a single buffer that is shared across all checkpoints, getting gradually filled in. To support this case then we actually have a special argument `eqxi.while_loop(..., buffers=...)`, to declare which elements of our loop state have this behaviour. Unfortunately that's not the case here because we do overwrite the values. (And side-note the presence of this `buffers` parameter is the reason I've not made this public API in Equinox, because the buffer-ness is completely unchecked and it's very easy to shoot yourself in the foot.)

andyElking replied:
Okay, I see. Thanks for the in-depth explanation! So let's see if I understand this correctly:

- I should set the default value to be 100 and keep the type as `Optional[int]`.
- If this was not getting rewritten, then I should make it register as a buffer in the outer while loop of diffeqsolve. But, because it does get rewritten, I should not do that(??). Still, I am curious, if I did want to register it as a buffer, how would I accomplish that? Is it indeed by making it an instance of `SaveState`, or is it something else entirely?

patrick-kidger commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,410 @@
+                jump_keep, eqxi.nextafter(eqxi.nextafter(next_t0)), next_t0
+            )
+
+        if TYPE_CHECKING:
+            assert isinstance(
+                next_t0, RealScalarLike
+            ), f"type(next_t0) = {type(next_t0)}"
```
I think we should be able to do this check even outside of type checking?

andyElking replied:
I get this:
```
AssertionError: type(next_t0) = <class 'jax._src.interpreters.partial_eval.DynamicJaxprTracer'>
```

patrick-kidger replied:
Oh, this is a silly known Python bug. You need `isinstance(next_to, typing.get_args(RealScalarLike))`. Python's unions have the wrong `isinstance` behaviour.

andyElking commented:
Thanks for the review! I made all the edits I could and I left some comments where I need guidance (no hurry though, this is not high priority for me). Also, should I get rid of `prev_dt` entirely, as you suggested in #483?

andyElking commented:
Working on the NeuralSDE now and I realised JumpStepWrapper can be useful to wrap a constant step controller as well, for the purpose of making it step to save_ts on top of whichever constant steps it is making. Yes it is more efficient/proper to just do a union of the original constant steps and the `save_ts` and then feed that into diffrax.StepTo. But there are situations where that could be very inconvenient to do in code, so you can just wrap your constant step controller with a `JumpStepWrapper`. And I am pretty sure the overhead should be quite minimal (with my implementation the cost per step is no longer O(log n), but just O(1)), so this isn't so bad.

This ties into the SDE interpolation business, where essentially the only good way of interpolating SDEs seems to be just making the solver step to whichever ts you want to save at. I do intend to eventually implement that, but in the meantime this is a useful workaround, I would say.

We don't really have to advertise this use case, but can nevertheless just change the type of `JumpStepWrapper.controller` from `AbstractAdaptiveStepController` to `AbstractStepController`.
What do you think?

patrick-kidger commented:
> Also, should I get rid of prev_dt entirely, as you suggested in #483?

If it's easy to do that in a separate commit afterwards then I would say yes. A separate commit just so it's easy to revert if it turns out we were wrong about something here :D

patrick-kidger commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,424 @@
+        if step_ts is not None:
+            # If we stepped to `t1 == step_ts[i_step]` and kept the step, then we
+            # increment i_step and move on to the next t in step_ts.
+            step_inc_cond = keep_step & (t1 == _get_t(i_step, step_ts))
```
I don't think I'm comfortable with this `==` between floating point numbers.

More generally speaking I think there could be cases in which the times being passed here do not perfectly align with the times that the adaptive step size controller suggested on the previous step (e.g. because of further wrapping of the step size controller), so I think this kind of logic is wrong anyway. I think you need something more like the `jump_ts` branch below, where you just want to snap `i_step` to the correct value. (Nothing that the correct value should be determinable statically, we only have state here for efficiency purposes.)

andyElking replied:
I implemented a linear search as you suggest and now use it to determine `i_step` and `i_jump`.

However, I think that for the rejected step business using linear search is incorrect, because `t1` should never be greater than `rejected_t = _get_t(i_reject, rejected_buffer)`. So depending on what you think is more appropriate we can either use `==` or  `jnp.isclose(t1, rejected_t, atol=1e-12)`. I also added the code below, but I can remove it if you don't think it is necessary. I could also add a parameter that only activates this callback when we are doing tests. Let me know.

``` python
i_reject = eqx.error_if(
    i_reject,
    t1 > rejected_t,
    "Jumped over a rejected time. Please report this as a bug.",
)
```

patrick-kidger replied:
FWIW if you want `error_if` here then I'd make it a testing-only branch. During regular user runtime I try to avoid using it for these kinds of asserts as it's actually quite slow.

patrick-kidger commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,424 @@
+            i_step = jnp.where(step_inc_cond, i_step + 1, i_step)
+
+        if jump_ts is not None:
+            next_jump_t = _get_t(i_jump, jump_ts)
+            jump_inc_cond = keep_step & (t1 >= eqxi.prevbefore(next_jump_t))
+            i_jump = jnp.where(jump_inc_cond, i_jump + 1, i_jump)
```
Likewise, if someone else is fiddling with the times then this seems to me like it might be fragile.

Recalling the previous implementation with its inefficient use of searchsorted. I think the robust approach here might be to write something with the same API as that, but whose implementation is just a simple linear search forwards or backwards from the current position (which is a 'hint' about where to start searching).

Most of the time that will just iterate once and be done, as here. But in the edge cases it should now do the right thing.

andyElking commented:
Thanks for the review, Patrick! I'll probably make the fixes sometime in the coming week. I am also making progress on the ML examples for the Single-seed paper, but it is slower now, due to my internship.

andyElking commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,424 @@
+    return _t1
+
+
+# EXPLANATION OF STEP_TS AND JUMP_TS
```
I moved this entire explanation into the docstring. Not sure if you intended me to move it all, but I can easily move it back. I also rewrote the little `!!! warning` in `stepsize_controller.md` about what requirements need to be fulfilled in order for an adaptive SDE solve to be guaranteed to converge to the correct solution.

andyElking commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,424 @@
+    jump_ts: Optional[Array]
+    inner_state: _ControllerState
+
+    def get(self):
```
That makes sense. I did it mostly for convenience, but it's better to stick with the good practice.

andyElking commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,424 @@
+            keep_step,
+            next_t0,
+            next_t1,
+            _,
```
I've given this a hard think and I think it is nearly impossible to do this perfectly. We can maintain correctness of the DE solution by allowing `made_jump` to sometimes be a false positive (e.g. the inner JSW recorded `jump_next_step=True` in the previous step, but then the outer JSW further clipped the proposal, so the jump didn't actually happen). That just makes us do one unnecessary VF evaluation at the fanthom jump point, but if I understand correctly the final solution should still be correct.

andyElking replied:
I handled this all inside the JSW in the new commit and I wrote a long comment about why I think the implementation is correct. But please do double check it.

andyElking commented:
I am very confused about what the correct value of `made_jump` should be when the step was rejected. By my understanding the original PID controller also got this wrong. However, I'm not sure whether getting this wrong is actually a significant issue, so I want your thoughts on it.

Suppose there is a jump at t=2. I will present 2 possible scenarios, in both of which I think something goes wrong (although maybe diffeqsolve might correct for the issue in scenario B). I wrote them as if JSW and the controller are separate, but the same holds for just the old PID controller.

====== scenario A =======

1. We start with a step from `t0 = 0` to `t1 = 1`, the controller decides the step will be kept and makes a proposal `next_t0 = 1`, `next_t1=2.5` which gets clipped to `next_t1 = 2` and so the JSW writes `jump_next_step=True` in its state.
2. We take the next step from `t0 = 1` to `t1 = 2`. Suppose the step is rejected, so `next_t0 = 1`. But the way the code works now, it will still set `made_jump = previous_state.jump_next_step`, which has been set to `True`. But that is wrong because `made_jump` should reflect whether there is a jump at `next_t0` so that the solver knows whether to reevaluate the VF at the start of the next step. Still this is not horrible, since it just causes one extra evaluation of the VF, but the full solution should still be correct. Right?

====== scenario B =======

1. We start with a step from `t0 = 0` to `t1 = 1`, the controller decides the step will be kept and makes a proposal `next_t0 = 1`, `next_t1=2.5` which gets clipped to `next_t1 = 2` and so the JSW writes `jump_next_step=True` in its state.
2. We take the next step from `t0 = 1` to `t1 = 2`. Suppose the step is accepted and the controller proposes `next_t0 = 2` and `next_t1 = 4` the controller returns (correctly) `made_jump = previous_state.jump_next_step (=True)` and `ControllerState(..., jump_next_step=False)`.
3. We take the next step from `t0 = 2` to `t1 = 4`. Note that because `made_jump` was set to `True` last iteration, the FSAL VF was reevaluated at `t = 2`. Suppose the step is rejected and the new proposal is `next_t0 = 2` and `next_t1 = 3`. The controller returns  `made_jump = previous_state.jump_next_step (=False)`. This is a problem because there is a jump at `next_t0=2`, so `made_jump` should have been `True`. However that might be fine if the value of the VF which was recomputed at the start of this step is kept for the next step (but I remember thinking that solver state is discarded when the step is rejected, so the recomputed VF also gets discarded).


Another way of seeing this all is through this:

- `jump_next_step = (there is a jump at next_t1)`
- `previous_state.jump_next_step = (there is a jump at t1)`
- `made_jump` should be `True` iff there is a jump at `next_t0`

Hence setting `made_jump = previous_state.jump_next_step` only works when the current step is being accepted, so `next_t0 = t1`. When rejecting the step, the new `made_jump` should be equal to the previous step's `made_jump` which we no longer have access to. The solution would be to add yet another thing to the state, but first I wanted to confirm with you to see if I'm getting this right. What are your thoughts?

patrick-kidger commented:
So I think the `made_jump`-on-rejected-step is handled through this line, outside the stepsize controller:

https://github.com/patrick-kidger/diffrax/blob/daec89c8fc9e72e4c2b32def2527a96b80f18b9d/diffrax/_integrate.py#L390

I made the decision to handle some of the step-rejection logic in the main `integrate.py` loop, on the basis that for those pieces it should be the same for all stepsize controllers.

So I *think* this fine? Do double-check my logic though! :p

Other than that, one thing I am noticing is that this `next_made_jump` business is pretty annoying to deal with in the stepsize controller, and I think could probably also be factored out to happen in the main `integrate.py` loop.

andyElking commented:
Great, that's exactly the line I was looking for (I must admit I looked in `_integrate.py` only briefly). But yes this is precisely the piece of the puzzle that my logic was missing, so I think we both independently arrived at the same conclustion that this is probably correct.

Thinking about it now, the `next_made_jump` situation might be way easier than I thought it was. I'll try to write the code and then I'll write a proof that it does the right thing for any number of stacked JSWs.

andyElking commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,424 @@
+
+def _get_t(i: IntScalarLike, ts: Array) -> RealScalarLike:
+    i_min_len = jnp.minimum(i, len(ts) - 1)
+    return jnp.where(i == len(ts), jnp.inf, ts[i_min_len])
```
For each of the tests (except backprop) in `test_adaptive_stepsize_controller.py` I added `@pytest.mark.parametrize("backwards", [False, True])` which just swaps the order of `t0` and `t1`. This also revealed two other problems:

1.  `step_ts` and `jump_ts` need to be sorted and in particular need to be re-sorted after multiplying them with `direction` in `wrap`. I added sorting both in `JSW.__init__` and `JSW.wrap`.
2. VBT complains if `t0 >= t1`. I don't think this is necessarily a problem but it could be confusing to some users, so let me know if you want to revisit that design decision and I can try to do something about it.

But other than that it all seems to work perfectly when solving backwards. To be fair the way you dealt with the whole backwards solve business is very clean and makes this work perfectly without any alterations.

patrick-kidger replied:
Okay, great! Then I'm very glad to have these tests :D 

For VBT I think I'm happy with either approach -- autoswitching `t0`, `t1` or just raising an error. No strong feelings on what is the better UX.

andyElking commented on diffrax/_step_size_controller/pid.py:
```diff
@@ -600,10 +542,12 @@ def _scale(_y0, _y1_candidate, _y_error):
         factor3 = 1 if _zero_coeff(coeff3) else prev_prev_inv_scaled_error**coeff3
         factormin = jnp.where(keep_step, 1, self.factormin)
+        # If the step is not kept, next step must be smaller, so factor must be <1.
+        factormax = jnp.where(keep_step, self.factormax, self.safety)
```
Just pushed a rebased version of all this, where these changes are in separate commits.

andyElking commented:
Hi Patrick!

I just pushed a new version of this PR, rebased on top of the most current main. I think I addressed everything you asked me to fix.

As it stands this contains 3 commits, contatining:

1. Everything except 2. and 3.
2. The changes to `at_dtmin` and `factormax` in `pid.py`
3. Removing `prev_dt` from JSW.

I left some conversations unresloved. I did try to fix the things mentioned in those, but I am not sure whether what I did was the best way to tackle that so I wanted to hear your opinion.

Also the test are failing because pyright doesn't know how to import `typeguard`, which has nothing to do with my changes.

patrick-kidger commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,463 @@
+        next_t1, _ = _clip_ts(next_t0, next_t1, i_step, st.step_ts, False)
+        next_t1, jump_at_next_t1 = _clip_ts(next_t0, next_t1, i_jump, st.jump_ts, True)
+
+        # Let's prove that the line below is correct. Say the inner controller is
+        # itself a JumpStepWrapper (JSW) with some inner_jump_ts. Then, given that
+        # it propsed (next_t0, original_next_t1), there cannot be any jumps in
+        # inner_jump_ts between next_t0 and original_next_t1. So if the next_t1
+        # proposed by the outer JSW is different from the original_next_t1 then
+        # next_t1 \in (next_t0, original_next_t1) and hence there cannot be a jump
+        # in inner_jump_ts at next_t1. So the jump_at_next_t1 only depends on
+        # jump_at_next_t1.
+        # On the other hand if original_next_t1 == next_t1, then we just take an
+        # OR of the two.
+        jump_at_next_t1 = jnp.where(
+            next_t1 == original_next_t1,
+            jump_at_original_next_t1,
+            jump_at_next_t1 | jump_at_original_next_t1,
+        )
```
Hmm, I don't think I completely believe this. Can we have the following:
- the PID controller proposes `t1`.
- the inner JSW wants to clip to a jump `b < t1`.
- the outer JSW wants to clip to a step (not a jump!) `a < b`
?

In this case then we will have `next_t1 != original_next_t1`, an `jump_at_original_next_t1 == True`... but overall we want `made_jump == False`?

patrick-kidger replied:
+can we have a test for two tested JSW, including the above scenario? It doesn't need to be a full diffeqsolve, just directly calling `adapt_step_size` and checking that we get the right output.

andyElking replied:
I think I bungled the code completely, because it doesn't do what's written in the comment at all. But I believe the comment (if implemented correctly) is correct.

I think the code should be:

```
jump_at_next_t1 = jnp.where(
    next_t1 == original_next_t1,
    jump_at_next_t1 | jump_at_original_next_t1,
    jump_at_next_t1,
)
```

And this indeed works in the case you brought up as well.

Given this, I don't think moving `next_made_jump` into integrate is strictly necessary at this point.

andyElking replied:
Hmm, yes I'll add a test.

patrick-kidger reviewed (commented):
Okay! I think I really like this.

First of all, I think I'm basically happy with pretty much everything outside of `jump_step_wrapper.py`. The changes here are pleasingly simple ^^

For `jump_step_wrapper.py`, I think my main question is around whether the rejected-step-buffer should actually be part of this wrapper at all -- since that handles SDEs with any kind of step rejection, which I think is completely orthogonal to clipping steps? (Not sure how I didn't notice this before!) I've also commented on a few other more minor points.

By the way, what did you think of the idea of moving `next_made_jump` into `_integrate.py`? It doesn't have to be now -- happy for that to be a separate PR -- just checking your thoughts on whether it is a generalisable thing.

Finally: merry Christmas, and a happy new year! :D

andyElking commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,463 @@
+
+
+_ControllerState = TypeVar("_ControllerState")
+_Dt0 = TypeVar("_Dt0", None, RealScalarLike, Optional[RealScalarLike])
```
Oh, I see, thanks for this explanation! I remember getting confused about this before, and I think I looked it up more than once, but I never got such a good explanation!

andyElking commented on diffrax/_step_size_controller/pid.py:
```diff
@@ -576,7 +510,9 @@ def _scale(_y0, _y1_candidate, _y_error):
         keep_step = scaled_error < 1
+        # Automatically keep the step if we're at dtmin.
         if self.dtmin is not None:
+            at_dtmin = at_dtmin | (prev_dt <= self.dtmin)
             keep_step = keep_step | at_dtmin
```
Yes, good point. Looking at the code, I don't really see a reason to keep it, but I'll get rid of it in a separate commit just to be sure.

andyElking commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,463 @@
+
+    Explanation of `step_ts` and `jump_ts`:
+
+    The `step_ts` and `jump_ts` are used to force the solver to step to certain times.
+    They mostly act in the same way, except that when we hit an element of `jump_ts`,
+    the controller must return `made_jump = True`, so that the diffeqsolve function
+    knows that the vector field has a discontinuity at that point, in which case it
+    re-evaluates it right after the jump point. In addition, the
+    exact time of the jump will be skipped using eqxi.prevbefore and eqxi.nextafter.
+    So now to the explanation of the two (we will use `step_ts` as an example, but the
+    same applies to `jump_ts`):
```
Yes, I intended to keep that as a comment, not part of the docs, but you suggested I put it in the docstring and I wasn't sure what exactly you wanted. I don't have strong opinions here, so feel free to rewrite it however you wish.

andyElking commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,463 @@
+    We use a convenience function _get_t(i, ts) which returns ts[i] if i < len(ts) and
+    infinity otherwise.
+
+    Explanation of revisiting rejected steps:
+
+    This feature should be used if and only if solving SDEs with non-commutative noise
+    using an adaptive step controller.
+
+    We use a "stack" of rejected steps, composed of a buffer `rejected_buffer` of length
+    `rejected_step_buffer_len` and a counter `i_reject`. The "stack" are all the items
+    in `rejected_buffer[i_reject:]` with `rejected_buffer[i_reject]` being the top of
+    the stack.
+    When `i_reject == rejected_step_buffer_len`, the stack is empty.
+    At the start of the run, `i_reject = rejected_step_buffer_len`. Each time a step is
+    rejected `i_reject -=1` and `rejected_buffer[i_reject] = t1`. Each time a step ends
+    at `t1 == rejected_buffer[i_reject]`, we increment `i_reject` by 1 (even if the
+    step was rejected, in which case we will re-add `t1` to the stack immediately).
+    We clip the next step to `t1_next = min(t1_next, rejected_buffer[i_reject])`.
+    If `i_reject < 0` then an error is raised.
```
Hmm I see what you mean in the first comment. So essentially as an attribute that user would have to manually set to `True`? That could probably work. Then again, we did put up several warnings and beyond that we do have to assume the user isn't stupid anyway. But if you would like me to implement such a thing, I can try.

For answer to second comment see my more in-depth response below.

patrick-kidger replied:
> Hmm I see what you mean in the first comment. So essentially as an attribute that user would have to manually set to True?

Yup exactly. Just as an extra preventative footgun.

andyElking commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,463 @@
+        return (_i > 0) & (ts[_i - 1] > t)
+
+    i = hint
+    i = jax.lax.while_loop(cond_up, lambda _i: _i + 1, i)
+    i = jax.lax.while_loop(cond_down, lambda _i: _i - 1, i)
```
At most one of the two loops will trigger, so I am pretty sure doing it this way is faster and cleaner than your second suggestion (unless I'm getting it completely wrong??). And this is probably the safest option, but yeah I think we can easily just have the upwards loop, everything should still work if my logic is correct. Well in fact everything worked perfectly without any loops at all, the reason we added this is to be extra sure there aren't any edge cases. So I'd say if we want to be safe, let's be completely safe and have both loops. In my experiments the slowdown from this is relatively small.

andyElking commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,463 @@
+
+        # This is just a logging utility for testing purposes
+        if self.callback_on_reject is not None:
+            jax.debug.callback(self.callback_on_reject, keep_step, t1)
```
Sounds good! I don't think the order matters in the test I use this for, but I suppose might as well do it properly.

andyElking commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,463 @@
+                next_t0, RealScalarLike
+            ), f"type(next_t0) = {type(next_t0)}"
+        else:
+            isinstance(next_t0, get_args(RealScalarLike))
```
Haha oops

andyElking commented on diffrax/_step_size_controller/jump_step_wrapper.py:
```diff
@@ -0,0 +1,463 @@
+        if TYPE_CHECKING:  # if i don't seperate this out pyright complains
+            assert isinstance(
+                next_t0, RealScalarLike
+            ), f"type(next_t0) = {type(next_t0)}"
```
Oh, I see.

andyElking commented:
Hi Patrick!

Sorry for the long silence, the last few weeks have been very busy.

I made the corrections you suggested.

Regarding splitting off revisiting steps from `JumpStepWrapper`, I agree that this feature is perpendicular to `jump_ts` and `step_ts`, but I don't see the harm in implementing several different features with one class... Certainly we want step-revisiting to be part of some sort of wrapper which can wrap either PID or potentially a custom step controller (I had such a situation in the paper with James I think), so it shouldn't be part of PID either.
So then the alternative is to have two wrappers -- `JumpStepWrapper` and `RevisitingWrapper`  -- which might work, but might also introduce more confusion and more potential edge cases.

So unless you feel very strongly about separating these two features, I would prefer not to add extra hurdles to this PR and conclude it in the near future. I think Owen has been itching to get this done as fast as possible as well.

patrick-kidger commented:
> So unless you feel very strongly about separating these two features, I would prefer not to add extra hurdles to this PR and conclude it in the near future. I think Owen has been itching to get this done as fast as possible as well.

Leaving it here on that basis sounds totally reasonable to me. Let's just tweak the name from `JumpStepWrapper` to something else to reflect this though.

patrick-kidger commented on test/test_adaptive_stepsize_controller.py:
```diff
@@ -65,6 +75,7 @@ def run(**kwargs):
     sol_no_jump_ts = run()
     sol_with_jump_ts = run(jump_ts=[7.5])
     assert sol_no_jump_ts.stats["num_steps"] > sol_with_jump_ts.stats["num_steps"]
+    print(sol_no_jump_ts.stats["num_steps"], sol_with_jump_ts.stats["num_steps"])
```
!

patrick-kidger reviewed (commented):
Okay, I started reviewing this... and now it's way too late in the evening for me to parse the math on this! :D 
The one big thing I want to check for is that the numerics in `JumpStepWrapper` are doing the correct thing. I'll do that tomorrow and then let's aim to merge this. :)

patrick-kidger commented:
Okay @andyElking, I've finished going over this PR and I've pushed changes back to this branch in one commit. Changes are:

- Renamed `JumpStepWrapper` to `ClipStepSizeController` to reflect the three different kinds of clipping it does.
- Simplify the internals such that `st.jump_at_next_t1` does not need to be stored: instead I examine the previous times we were stepping to. I think this should be much simpler to reason about! (The previous implementation here / in `PIDController` was always something I found pretty mind-bending.) More pragmatically, I'm more certain that this will do the right thing when nesting multiple wrappers inside each other.
- Switched from using `eqx.error_if` to returning an error code, which should be more efficient.
- Adjusted documentation and added your paper to `diffrax.citation` :)

Lmk what you think!

andyElking commented:
I read your changes and it all LGTM. Indeed the trick of just checking equality over all `ts` in `_bump_next_t0` is very inventive and much cleaner! You really made the whole code waaay simpler!

Oh, I never really thought of using the `RESULT` business to report errors. To be fair, now it makes much more sense to me why it's there in the first place.

I completely agree with the renaming as well.

Thanks for adding the citation!

Also I find the `_set_metaclass = dict(metaclass=_MetaPID)` trick pretty funny. Crazy what sorts of workarounds are sometimes needed in python 😁

Thanks for finishing this up!

patrick-kidger commented:
Awesome. In that case: now merged :) Thank you for your efforts driving this one!
